File: DMOJ/graphtheory/bts18p4.py
from collections import defaultdict, deque
import math
good = [False for _ in range(100010)]
visited1 = [False for _ in range(100010)]
visited2 = [False for _ in range(100010)]
graph = defaultdict(list)

firstDist, firstNode, secondDist, secondNode = -1, -1, -1, -1
maxDist, maxNode = 0, 0

n = int(input())
strengths = list(map(int, input().split()))
for i in range(1, len(strengths) + 1):
    strength = strengths[i - 1]
    dis = 1 + 4 * 1 * strength
    if int(math.sqrt(dis)) == math.sqrt(dis) and int(math.sqrt(dis)) % 2 == 1:
        good[i] = True

for i in range(n - 1):
    a, b = map(int, input().split())
    if good[a] and good[b]:
        graph[a].append(b)
        graph[b].append(a)

# print(*good[0:n + 1])

for i in range(1, n + 1):
    if visited1[i] or not good[i]: continue
    q = deque([(i, 1)])
    visited1[i] = True
    # print("start")
    while q:
        cn, cd = q.popleft()
        # print(f"{cd = } {cn = }")
        for e in graph[cn]:
            if visited1[e]: continue
            visited1[e] = True
            q.append((e, cd + 1))

    q = deque([(cn, 1)])
    visited2[cn] = True
    while q:
        cn, cd = q.popleft()
        # print(f"{cd = } {cn = }")
        for e in graph[cn]:
            if visited2[e]: continue
            visited2[e] = True
            q.append((e, cd + 1))
    
    if (cd >= maxDist):
        maxNode = cn
        maxDist = cd


print(f"{maxDist}")

File: DMOJ/graphtheory/bts18p4.cpp
#include <bits/stdc++.h>

using std::cin;
using std::unordered_map;
using std::vector;

typedef long long ll;
typedef long double ld;
int n;
bool good[100010];
bool visited[100010];
ll strength;
unordered_map<int, vector<int>> graph;

int firstNode, secondNode, firstDist, secondDist;
int maxDist, maxNode;

void dfs(int node, int dist, int past, int type)
{
    // printf("a\n");
    visited[node] = true;
    if (type == 1 && dist > firstDist)
    {
        firstDist = dist;
        firstNode = node;
    }

    if (type == 2 && dist > secondDist)
    {
        secondDist = dist;
        secondNode = node;
    }

    for (int edge : graph[node])
    {
        if (edge == past)
            continue;
        dfs(edge, dist + 1, node, type);
    }
}

int main()
{
    cin >> n;
    // printf("%lld\n", sqrtl(ld(10000000000000000)));
    for (int i = 1; i <= n; i++)
    {
        cin >> strength;
        ld dis = 1 + 4 * 1 * strength;
        if (ll(sqrtl(dis)) == sqrtl(dis) && ll(sqrtl(dis)) % 2 == 1)
        {
            good[i] = true;
        }
    }
    // printf("a");

    for (int i = 0, a, b; i < n - 1; i++)
    {
        cin >> a >> b;
        if (good[a] && good[b])
        {
            graph[a].push_back(b);
            graph[b].push_back(a);
        }
    }
    // printf("b");

    for (int i = 1; i <= n; i++)
    {
        if (visited[i] || !good[i])
            continue;
        dfs(i, 1, -1, 1);
        dfs(firstNode, 1, -1, 2);
        if (secondDist >= maxDist)
        {
            maxNode = secondNode;
            maxDist = secondDist;
        }
    }

    printf("%d\n", maxDist);

    return 0;
}